### PR TITLE
[full-ci][tests-only] Extend tests coverage for adding users in a group with single request

### PR DESCRIPTION
**Pull request by amrita-shrestha**
## Description
This PR adds tests related to adding users in a group with single request

## Related Issue
- https://github.com/owncloud/ocis/issues/5569

## How Has This Been Tested?

- Locally

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: tests/TestHelpers/GraphHelper.php [hunk 1]
```diff
@@ -532,6 +550,39 @@ public static function addUserToGroup(
...
+	 * @param string $xRequestId
+	 * @param string $adminUser
+	 * @param string $adminPassword
+	 * @param array $usersId
+	 * @param string $groupId
+	 *
+	 * @return ResponseInterface
+	 * @throws GuzzleException
+	 */
+	public static function addUsersToGroupInASingleRequest(
```
Comment (saw-jan) [inline tests/TestHelpers/GraphHelper.php:564]:
```suggestion
	public static function addMultipleUsersToGroup(
```

File: tests/TestHelpers/GraphHelper.php [hunk 2]
```diff
@@ -499,6 +499,24 @@ public static function deleteGroup(
...
+	public static function createBodyForUsersAdditionInGroupBySingleRequest(string $baseUrl, array $usersId): array {
+		$fullUrl = [];
+		foreach ($usersId as $userId) {
+			$fullUrl[] = self::getFullUrl($baseUrl, 'users/' . $userId);
+		}
+		return [
+			"members@odata.bind" => $fullUrl
+		];
+	}
+
```
Comment (saw-jan) [inline tests/TestHelpers/GraphHelper.php:519]:
```suggestion
```

File: tests/TestHelpers/GraphHelper.php [hunk 3]
```diff
@@ -532,6 +550,39 @@ public static function addUserToGroup(
...
+	public static function addUsersToGroupInASingleRequest(
+		string $baseUrl,
+		string $xRequestId,
+		string $adminUser,
+		string $adminPassword,
+		array $usersId,
+		string $groupId
+	): ResponseInterface {
+		$url = self::getFullUrl($baseUrl, 'groups/' . $groupId);
+		$body = \json_encode(self::createBodyForUsersAdditionInGroupBySingleRequest($baseUrl, $usersId));
```
Comment (saw-jan) [inline tests/TestHelpers/GraphHelper.php:573]:
```suggestion
                $body = [ "members@odata.bind" => [] ];
		foreach ($usersId as $userId) {
			$body["members@odata.bind"][] = self::getFullUrl($baseUrl, 'users/' . $userId);
		}
		$body = \json_encode($body);
```

**Review by saw-jan — COMMENTED** [review]
should be fine to add it to the same function. Do we need it elsewhere?

File: tests/TestHelpers/GraphHelper.php [hunk 4]
```diff
@@ -532,6 +550,39 @@ public static function addUserToGroup(
...
+	 *
+	 * @return ResponseInterface
+	 * @throws GuzzleException
+	 */
+	public static function addUsersToGroupInASingleRequest(
+		string $baseUrl,
+		string $xRequestId,
+		string $adminUser,
+		string $adminPassword,
+		array $usersId,
```
Comment (saw-jan) [inline tests/TestHelpers/GraphHelper.php:569]:
```suggestion
		array $userIds,
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 5]
```diff
@@ -1563,4 +1563,46 @@ public function userGetsAllUsersOfTwoGroupsUsingTheGraphApi(string $user, string
...
+	 * @param TableNode $table
+	 *
+	 * @return void
+	 * @throws Exception
+	 * @throws GuzzleException
+	 */
+	public function theAdministratorAddsTheFollowingUsersToAGroupInASingleRequestUsingTheGraphApi(string $user, string $group, TableNode $table): void {
+		$credentials = $this->getAdminOrUserCredentials($user);
+		$this->featureContext->verifyTableNodeColumns($table, ['username']);
+		$usersId = [];
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:1581]:
```suggestion
		$userIds = [];
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 6]
```diff
@@ -1563,4 +1563,46 @@ public function userGetsAllUsersOfTwoGroupsUsingTheGraphApi(string $user, string
...
+			$groupId = WebDavHelper::generateUUIDv4();
+		}
+
+		foreach ($table->getHash() as $row) {
+			try {
+				$usersId[] = $this->featureContext->getAttributeOfCreatedUser($row['username'], "id");
+			} catch (Exception $e) {
+				$usersId[] = WebDavHelper::generateUUIDv4();
+			}
+		}
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:1595]: Let the step only do the positive cases. If we need to test trying to add users to non-existing group then we can create another different step.
```suggestion
		$groupId = $this->featureContext->getAttributeOfCreatedGroup($group, "id");

		foreach ($table->getHash() as $row) {
			$usersId[] = $this->featureContext->getAttributeOfCreatedUser($row['username'], "id");
		}
```

File: tests/acceptance/features/apiGraph/addUserToGroup.feature [hunk 7]
```diff
@@ -149,3 +149,21 @@ Feature: add users to group
...
+
+
+  Scenario: add multiple users to a group in a single request
+    Given the administrator has given "Alice" the role "Admin" using the settings api
+    And these users have been created with default attributes and without skeleton files:
+      | username |
+      | Brian    |
+      | Carol    |
+    And user "Alice" creates a group "grp1" using the Graph API
+    When the administrator "Alice" adds the following users to a group "grp1" in a single request using the Graph API
```
Comment (SagarGi) [inline tests/acceptance/features/apiGraph/addUserToGroup.feature:161]:
```suggestion
    When the administrator "Alice" adds the following users to a group "grp1" using the Graph API
```

File: tests/TestHelpers/GraphHelper.php (hunk 2 above)
Comment (amrita-shrestha) [reply tests/TestHelpers/GraphHelper.php:519]: @saw-jan what u are suggesting?

File: tests/acceptance/features/apiGraph/addUserToGroup.feature (hunk 7 above)
Comment (amrita-shrestha) [reply tests/acceptance/features/apiGraph/addUserToGroup.feature:161]: Don't you think it will be nice if we differentiate single and multiple requests

Comment (saw-jan) [reply tests/acceptance/features/apiGraph/addUserToGroup.feature:161]: `at once` would be nice for readability.

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 6 above)
Comment (amrita-shrestha) [reply tests/acceptance/features/bootstrap/GraphContext.php:1595]: I was going to add a scenario where we need this code

File: tests/acceptance/features/apiGraph/addUserToGroup.feature [hunk 8]
```diff
@@ -149,3 +149,44 @@ Feature: add users to group
...
+      | Brian    |
+      | Carol    |
+    Then the HTTP status code should be "204"
+    And the following users should be listed in the following groups
+      | username | groupname |
+      | Brian    | grp1      |
+      | Carol    | grp1      |
+
+
+  Scenario: admin tries to add users at once to a group without sending the group
```
Comment (SagarGi) [inline tests/acceptance/features/apiGraph/addUserToGroup.feature:172]:
```suggestion
  Scenario: admin tries to add users at once to a non-existing group
```
you can change if it cab be like this ?

File: tests/acceptance/features/apiGraph/addUserToGroup.feature [hunk 9]
```diff
@@ -149,3 +149,44 @@ Feature: add users to group
...
+      | Brian    |
+      | Carol    |
+    When the administrator "Alice" adds the following users to a group "" at once using the Graph API
+      | username |
+      | Brian    |
+      | Carol    |
+    Then the HTTP status code should be "404"
+
+
+  Scenario: admin tries to add multiple non-existing at once user to a group
```
Comment (SagarGi) [inline tests/acceptance/features/apiGraph/addUserToGroup.feature:185]:
```suggestion
  Scenario: admin tries to add multiple non-existing user at once to a group
```

File: tests/acceptance/features/apiGraph/addUserToGroup.feature [hunk 10]
```diff
@@ -149,3 +149,44 @@ Feature: add users to group
...
+      | Carol    | grp1      |
+
+
+  Scenario: admin tries to add users at once to a group without sending the group
+    Given the administrator has given "Alice" the role "Admin" using the settings api
+    And these users have been created with default attributes and without skeleton files:
+      | username |
+      | Brian    |
+      | Carol    |
+    When the administrator "Alice" adds the following users to a group "" at once using the Graph API
```
Comment (grgprarup) [inline tests/acceptance/features/apiGraph/addUserToGroup.feature:178]: IMO better to create separate steps for negative cases. We can add examples for a non-existent group also.

File: tests/acceptance/features/apiGraph/addUserToGroup.feature (hunk 9 above)
Comment (grgprarup) [inline tests/acceptance/features/apiGraph/addUserToGroup.feature:185]:
```suggestion
  Scenario: admin tries to add multiple non-existing users at once to a group
  OR
  Scenario: admin tries to add multiple non-existing users to a group at once
```
A small question, what happens when a combination of existing and non-existing users is given? Would you be willing to add scenarios for such edge cases?

File: tests/acceptance/features/apiGraph/addUserToGroup.feature (hunk 10 above)
Comment (amrita-shrestha) [reply tests/acceptance/features/apiGraph/addUserToGroup.feature:178]: similar other steps are handled in the same way so its a topic of discussion. @saw-jan @SagarGi @saw-jan . If we make separate step then we have to refactor other step too which has been added in old PR

File: tests/TestHelpers/GraphHelper.php [hunk 11]
```diff
@@ -532,6 +532,42 @@ public static function addUserToGroup(
...
+			$url,
+			$xRequestId,
+			'PATCH',
+			$adminUser,
+			$adminPassword,
+			self::getRequestHeaders(),
+			\json_encode($body)
+		);
+	}
+
```
Comment (grgprarup) [inline tests/TestHelpers/GraphHelper.php:570]: We have `addUsersToGroup` in `GraphHelper` can't we use or refactor that?

Comment (amrita-shrestha) [reply tests/TestHelpers/GraphHelper.php:570]: I don't think it will be beneficial to reuse that function . We have to create new function to create body where we have to use if else to create different type of body for different request.

Comment (amrita-shrestha) [reply tests/TestHelpers/GraphHelper.php:570]: this function has been removed in past. Don't know why it has been removed.

File: tests/acceptance/features/apiGraph/addUserToGroup.feature (hunk 10 above)
Comment (saw-jan) [reply tests/acceptance/features/apiGraph/addUserToGroup.feature:178]: This is a newly added step. So, it would be far better if we can use separate step for this e.g.:
`When the administrator "Alice" tries to add the following users to a non-existing group at once using the Graph API`
This way, we don't need to try-catch or use conditions with a single-step def function. Also, it isolates the positive and negative case steps.

We can refactor the already-added ones later.

Comment (amrita-shrestha) [reply tests/acceptance/features/apiGraph/addUserToGroup.feature:178]: I had create an issue to work on the refactor https://github.com/owncloud/ocis/issues/5646

**Comment by amrita-shrestha** [comment]
Will work on new PR
